pymc/logprob/mixture.py

```diff
@@ -483,3 +484,71 @@ def logprob_MixtureRV(
     "basic",
     "mixture",
 )
+
+
+class MeasurableIfElse(IfElse):
+    """Measurable subclass of IfElse operator."""
+
+
+MeasurableVariable.register(MeasurableIfElse)
+
+
+@node_rewriter([IfElse])
+def find_measurable_ifelse_mixture(fgraph, node):
+    rv_map_feature = getattr(fgraph, "preserve_rv_mappings", None)
+
+    if rv_map_feature is None:
+        return None  # pragma: no cover
+
+    if isinstance(node.op, MeasurableIfElse):
+        return None
+
+    # Check if all components are unvalued measuarable variables
+    if_var, *base_rvs = node.inputs
+
+    if not all(
+        (
+            rv.owner is not None
+            and isinstance(rv.owner.op, MeasurableVariable)
+            and rv not in rv_map_feature.rv_values
+        )
+        for rv in base_rvs
+    ):
+        return None  # pragma: no cover
+
+    unmeasurable_base_rvs = ignore_logprob_multiple_vars(base_rvs, rv_map_feature.rv_values)
+
+    return MeasurableIfElse(n_outs=node.op.n_outs).make_node(if_var, *unmeasurable_base_rvs).outputs
+
+
+measurable_ir_rewrites_db.register(
+    "find_measurable_ifelse_mixture",
+    find_measurable_ifelse_mixture,
+    "basic",
+    "mixture",
+)
+
+
+@_logprob.register(MeasurableIfElse)
+def logprob_ifelse(op, values, if_var, *base_rvs, **kwargs):
+    """Compute the log-likelihood graph for an `IfElse`."""
+    from pymc.pytensorf import replace_rvs_by_values
+
+    assert len(values) * 2 == len(base_rvs)
+
+    rvs_to_values_then = {then_rv: value for then_rv, value in zip(base_rvs[: len(values)], values)}
+    rvs_to_values_else = {else_rv: value for else_rv, value in zip(base_rvs[len(values) :], values)}
+
+    logps_then = [
+        logprob(rv_then, value, **kwargs) for rv_then, value in rvs_to_values_then.items()
```

Comment:
What’s the behaviour of logprob when the value variable doesn’t have the expected shape? Will it raise a ValueError? I think that would be bad in this case. Imagine if you have a step method on the condition variable. The stepper might choose to move the condition to an infeasible value and that would kill the sampling process. I would like the condition that doesn’t match shapes to simply return -inf logprob. That way the stepper would discard the proposal and stay in reasonable regions.

Reply:
The behavior of the logprob when the value variable doesn't match shape is exactly the same as the logprob of the underlying components. If it doesn't match, it will try to broadcast and fail if it cannot. Other than that we don't use size information in any of the core logprob functions.

`pm.logp(pm.Normal.dist(size=(4,)), np.ones((2,)))` will be happy to return a logp with two values.

One other case where this shows up is in graphs of the form `pt.ones((5,)) + pm.Normal.dist()` which we infer to have an equivalent logp as that of `pm.Normal.dist(shape=(5,))` even though the generative process contains only one true random variable, and not 5).

I think a bigger discussion needs to be had about the role of shape information in the random and logp graphs, so I wouldn't treat `IfElse` differently for now.

Reply:
Opened an issue here: https://github.com/pymc-devs/pymc-experimental/blob/main/pymc_experimental/marginal_model.py